src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php

```diff
@@ -190,124 +168,88 @@ public function handleError($type, $msg, $file, $line, $context = [])
+        register_shutdown_function(function () use ($isFailing, $groups, $configuration) {
             foreach ($this->deprecations as $group => $arrayOrInt) {
                 if (0 < (\is_int($arrayOrInt) ? $arrayOrInt : \count($arrayOrInt))) {
                     echo "Shutdown-time deprecations:\n";
                     break;
                 }
             }
 
-            $this->displayDeprecations($groups, $mode);
+            $this->displayDeprecations($groups, $configuration);
 
-            if ($isFailing || self::MODE_WEAK !== $mode && $mode < $this->deprecations['unsilencedCount'] + $this->deprecations['remainingCount'] + $this->deprecations['otherCount']) {
+            if ($isFailing || !$configuration->tolerates($this->deprecations)) {
                 exit(1);
             }
         });
     }
 
-    private function getMode()
+    private function getConfiguration()
     {
-        if (false !== $this->resolvedMode) {
-            return $this->resolvedMode;
+        if (null !== $this->configuration) {
+            return $this->configuration;
         }
-
         if (false === $mode = $this->mode) {
             $mode = getenv('SYMFONY_DEPRECATIONS_HELPER');
         }
-
-        if (self::MODE_DISABLED !== $mode
-            && self::MODE_WEAK !== $mode
-            && self::MODE_WEAK_VENDORS !== $mode
-            && (!isset($mode[0]) || '/' !== $mode[0])
-        ) {
-            $mode = preg_match('/^[1-9][0-9]*$/', $mode) ? (int) $mode : 0;
+        if ('strict' === $mode) {
+            return $this->configuration = Configuration::inStrictMode();
         }
-
-        return $this->resolvedMode = $mode;
-    }
-
-    /**
-     * @param string $path
-     *
-     * @return bool
-     */
-    private static function inVendors($path)
-    {
-        /** @var string[] absolute paths to vendor directories */
-        static $vendors;
-
-        if (null === $vendors) {
-            foreach (get_declared_classes() as $class) {
-                if ('C' !== $class[0] || 0 !== strpos($class, 'ComposerAutoloaderInit')) {
-                    continue;
-                }
-
-                $r = new \ReflectionClass($class);
-                $v = \dirname(\dirname($r->getFileName()));
-
-                if (file_exists($v.'/composer/installed.json')) {
-                    $vendors[] = $v;
-                }
-            }
+        if (self::MODE_DISABLED === $mode) {
+            return $this->configuration = Configuration::inDisabledMode();
         }
-
-        $realPath = realpath($path);
-
-        if (false === $realPath && '-' !== $path && 'Standard input code' !== $path) {
-            return true;
+        if ('weak' === $mode) {
+            return $this->configuration = Configuration::inWeakMode();
+        }
+        if (self::MODE_WEAK_VENDORS === $mode) {
+            echo sprintf('Setting SYMFONY_DEPRECATIONS_HELPER to "%s" is deprecated in favor of "max[self]=0"', $mode).PHP_EOL;
+            exit(1);
```

Comment:
this is a BC break, as it will exit without running the tests. This should report a self deprecation, and return a configuration based on `max[self]=0`

Reply:
This is master, and @nicolas-grekas asked me to change my previous code, that was as you described, for this.

Reply:
I'll work on a fix for this one.